blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DAbstractMetricsRW.java

```diff
@@ -0,0 +1,324 @@
+                    locatorToFuturesMap.put(locator, existing);
+                }
+                existing.addAll(selectFutures);
+
+            } catch (CacheException ex) {
+                LOG.error(String.format("Error looking up locator %s in cache", locator), ex);
+            }
+        }
+        return resultSetsToMetricData(locatorToFuturesMap, locatorIOMap, granularity);
+    }
+
+    /**
+     * Given a locator, a data type, a rollup type and granularity, return the correct
+     * {@link com.rackspacecloud.blueflood.io.datastax.DAbstractMetricIO} object
+     *
+     * @param locatorIOMap the Locator -> DabstractMetricIO map to add the new returned object to as a side-effect
+     * @param locator
+     * @param dataType
+     * @param rollupType
+     * @param gran
+     * @return
+     */
+    private DAbstractMetricIO getIO( Map<Locator, DAbstractMetricIO> locatorIOMap,
+                                     Locator locator,
+                                     String dataType,
+                                     String rollupType,
+                                     Granularity gran ) {
+
+        DAbstractMetricIO io;
+
+        if( rollupType == null && gran == Granularity.FULL ) {
+
+            io = simpleIO;
+        }
+        else {
+
+            // find out the rollupType for this locator
+            RollupType rType = RollupType.fromString( rollupType );
+
+            // get the right PreaggregatedIO class that can process
+            // this rollupType
+            io = rollupTypeToIO.get(rType);
+        }
+
+
+        if (io == null) {
+            throw new IllegalArgumentException(String.format("dont know how to handle locator=%s with rollupType=%s and type=%s", locator, rollupType, dataType));
+        }
+
+        // put everything in a map of locator -> io so
+        // we can use em up later
+        locatorIOMap.put( locator, io );
+
+        return io;
+    }
+
+
+    /**
+     * Fetches a {@link com.rackspacecloud.blueflood.types.Points} object for a
+     * particular locator and rollupType from the specified column family and
+     * range
+     *
+     * @param locator
+     * @param rollupType
+     * @param range
+     * @param columnFamilyName
+     * @return
+     */
+    @Override
+    public Points getDataToRollup(final Locator locator,
+                                  RollupType rollupType,
+                                  Range range,
+                                  String columnFamilyName) throws IOException, CacheException {
+        Timer.Context ctx = Instrumentation.getReadTimerContext(columnFamilyName);
+        try {
+            // read the rollup object from the proper IO class
+            DAbstractMetricIO io;
+
+            if( columnFamilyName.equals( CassandraModel.CF_METRICS_FULL_NAME) ) {
+
+                io = simpleIO;
+            }
+            else
+                io = rollupTypeToIO.get(rollupType);
```

Comment:
Take a look at RollupRunnable where the old AstyanaxReader.getDataToRoll() was called:
https://github.com/rackerlabs/blueflood/blob/master/blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java#L143

We are going to have to call either DBasicMetricsRW or DPreaggregatedMetricsRW depending on the metrics RollupType. It seems to be that this ```if ( columnFamilyName.equals(CassandraModel.CF_METRICS_FULL_NAME) )``` should be in the DBasicMetricsRW, no? 

One idea is to have an abstract ```getIO()``` method that returns you the right IO object. Move all the ```rollupTypeToIO`` map stuff to the DPreaggregatedMetricsRW (since it only deals with Rollup). Therefore, this DAbstractMetricsRW abstract class don't need to do all these logic to determine which IO object is appropriate.